[FLINK-4389] Expose metrics to WebFrontend

zentol opened the pull request:
This PR exposes metrics to the Webfrontend, as proposed in [FLIP-7](https://cwiki.apache.org/confluence/display/FLINK/FLIP-7%3A+Expose+metrics+to+WebInterface).

This PR builds on-top of #2300, meaning that 2866f56 is not part of the PR.

I've split the implementation into 5 commits that implement
* the generation of a separate scope string for the WebInterface
* the MetricQueryService, a separate actor running on all Job-/TaskManagers whose main purpose is to create and return a dump of the metrics when queried to do so
* the MetricStore, a nested data structure used in the WebInterface to store transmitted metrics
* the MetricFetcher, which is used by the WebInterface to fetch metrics from Job-/TaskManagers
* various MetricsHandler classes, which handle REST calls requesting specific metrics

### MetricQueryService
The MetricQueryService is an actor running inside the MetricRegistry acting like an unscheduled reporter that is queried from the outside for a report. The MetricRegistry notifies it of added/removed metrics whereas the MetricFetcher sends report requests to the JM/TM which are then forwarded to the MetricQueryService, which answers directly to the MetricFetcher.

The report is one big `Object[]`, which contains for each metric
 1. the type of the metric, encoded as a byte (so that we know how many values are transmitted)
 2. the fully qualified metric name (based on the separate format)
 3. the value(s) of the metric (turned into Strings for Gauges)

### MetricStore
The MetricStore is a relatively simple nested data-structure that contains one HashMap<String, Object> for every JM/TM/job/task. Received metrics are added to these HashMaps based on the format string. There is only a single MetricStore instance in the WebInterface.

### MetricFetcher
The MetricFetcher initiates the transfer and cleanup of metrics. It contains the MetricStore instance, which is accessed by MetricHandlers. The fetching is only done when a handler asks for it, with a minimum duration of 10 seconds between updates. As such no fetching will be done if the metrics are not accessed with REST calls.

The fetching procedure can be summed up in pseudo-code as following:
```
fetch():
	askJobManagerForJobDetails()
		=> retain all metrics belonging to the given jobs
	askJobManagerForMetrics()
		=> add received metrics to MetricStore
	askJobManagerForRegisteredTaskManagers()
		=> retain all metrics belonging to registered task managers
		=> for each TaskManager:
			askTaskManagerForMetrics()
				=> add received metrics to MetricStore
```

### MetricsHandler
The MetricsHandlers deal with two requests:
* getAllAvailableMetrics - any REST request that does not have a `get` query parameter is treated as a request for all available metrics for a given JM/TM/job/task, denoted by the REST path. The reply will be a JSON array, for example: `[{"id":"metric_1"},{"id":"metric_2"}]`
* getMetricValues - the Webfrontend can request the values for several metrics by passing a comma-separated list of metric id's as the `get` query parameter. The reply will be a JSON array of id:value pairs, for example: `[{"id":"metric_1", "value":"4"}]` or an empty string if an error occurred.

tillrohrmann commented:
The test case `TaskManagerComponentsStartupShutdownTest.testComponentsStartupShutdown` fails on Travis.

zentol commented:
Fixed the failing test.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/AbstractMetricsHandler.java:
```diff
@@ -0,0 +1,124 @@
+ * Abstract request handler that returns a list of all available metrics or the values for a set of metrics.
+ *
+ * If the query parameters do not contain a "get" parameter the list of all metrics is returned.
+ * {@code {"available": [ { "name" : "X", "id" : "X" } ] } }
```
`"name" : "X"` won't be written, will it?

zentol replied:
That javadoc is a bit outdated, it should be `{@code [ { "id" : "X" ) ] }

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricFetcher.java:
```diff
@@ -0,0 +1,185 @@
+			synchronized (metrics) {
+				switch ((byte) data[x++]) {
+					case 0:
+					case 1:
```
Could we introduce some enums or so for these values. Maybe we could use the byte value as the ordinal of the corresponding enum value.

zentol replied:
we could use the constants defined in the MetricQueryService, or convert those to an enum.

tillrohrmann replied:
Yeah either that or sending the data in a more structured fashion as discussed below.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricFetcher.java:
```diff
@@ -0,0 +1,185 @@
+						metrics.add(histogramName + "_p95", data[x++]);
+						metrics.add(histogramName + "_p98", data[x++]);
+						metrics.add(histogramName + "_p99", data[x++]);
+						metrics.add(histogramName + "_p999", data[x++]);
```
This is probably quite error prone, because whenever the histogram is changed (e.g. order of serialized fields) it has to be reflected here as well. I think it would be better if we would have a serializer which does the serialization and deserialization in one place.

zentol replied:
sounds reasonable

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java:
```diff
@@ -0,0 +1,172 @@
+
+	JobManagerMetricStore jobManager = new JobManagerMetricStore();
+	Map<String, TaskManagerMetricStore> taskManagers = new HashMap<>();
+	Map<String, JobMetricStore> jobs = new HashMap<>();
```
Can these fields be final?

zentol replied:
yes

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java:
```diff
@@ -0,0 +1,172 @@
+		TaskMetricStore task;
+
+		try {
+			String[] components = name.split(":");
```
Do we filter `:` out in user defined metric names?

zentol replied:
yes, the MetricQueryService does that

tillrohrmann commented on flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/metrics/AbstractMetricHandlerTest.java:
```diff
@@ -0,0 +1,158 @@
+import static org.junit.Assert.fail;
+import static org.powermock.api.mockito.PowerMockito.mock;
+
+public class AbstractMetricHandlerTest {
```
`extends TestLogger` missing

tillrohrmann replied:
Should be added to all tests because this makes debugging with Travis logs considerably easier.

tillrohrmann commented on flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/metrics/AbstractMetricHandlerTest.java:
```diff
@@ -0,0 +1,158 @@
+
+public class AbstractMetricHandlerTest {
+	@Test
+	public void testHandleRequest() throws Exception {
```
A short description would be helpful

tillrohrmann commented on flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/metrics/AbstractMetricHandlerTest.java:
```diff
@@ -0,0 +1,158 @@
+		pathParams.put("jobid", "nonexistant");
+
+		try {
+			handler.handleRequest(pathParams, queryParams, null);
```
Maybe we could check that the return value is `""`.

zentol replied:
good idea

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricQueryService.java:
```diff
@@ -0,0 +1,180 @@
+	@Override
+	public void onReceive(Object message) throws Exception {
+		try {
+			if (message instanceof Tuple3) { // add metric
```
I think we should create proper message types instead of using a `Tuple3` here.

zentol replied:
agreed.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricQueryService.java:
```diff
@@ -0,0 +1,180 @@
+				} else if (metric instanceof Histogram) {
+					histograms.put((Histogram) metric, name);
+				}
+			} else if (message instanceof Metric) { // remove metric
```
The same applies to the remove metric message and the create dump. Both should get a proper message, e.g. `RemoveMetric` and `CreateDump`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistry.java:
```diff
@@ -144,6 +152,30 @@ public MetricRegistry(Configuration config) {
+	 * @return ActorRef to the MetricQueryService
+     */
+	public Option<ActorRef> getQueryService() {
+		return queryService == null
```
this is equivalent to `Option.apply(queryService)`. If queryService == null, then the apply call will return a `None`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistry.java:
```diff
@@ -207,6 +239,9 @@ public void register(Metric metric, String metricName, MetricGroup group) {
 				}
 			}
+			if (queryService != null) {
+				notifyOfAddedMetric(queryService, metric, metricName, (AbstractMetricGroup) group);
```
I think it's always good practice to add the class whose static method on calls in front of the static method. Otherwise you easily think that this method is somewhere defined in this class. Consequently, I would discourage static imports a little bit.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java:
```diff
@@ -54,13 +53,19 @@
  * These metrics simply do not get reported any more, when created on a closed group.
  */
-public abstract class AbstractMetricGroup implements MetricGroup {
+public abstract class AbstractMetricGroup<A extends AbstractMetricGroup> implements MetricGroup {
```
Type parameter description in the JavaDocs is missing

zentol replied:
I will address this in #2300

tillrohrmann replied:
Sounds good. Haven't exclude the #2300 commit from the review. My bad.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/ComponentMetricGroup.java:
```diff
@@ -33,19 +36,42 @@
  * exclude it (for continuity of the namespace across failure and recovery).
  */
-public abstract class ComponentMetricGroup extends AbstractMetricGroup {
+public abstract class ComponentMetricGroup<P extends AbstractMetricGroup> extends AbstractMetricGroup<P> {
```
Type parameter in the description is missing.

tillrohrmann replied:
Is this an internal component like the `JobMetricGroup`?

zentol replied:
I'm not sure what you mean, but `JobMetricGroup` inherits from `ComponentMetricGroup`.

tillrohrmann replied:
The `JobMetricGroup` is annotated as `internal`. I was just wondering whether we should do the same for this metric group and the others (to which it applies) as well.

zentol replied:
yes they should, the must've been removed by accident when we moved them to `flink-runtime`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskManagerJobMetricGroup.java:
```diff
@@ -61,9 +58,7 @@ public TaskManagerJobMetricGroup(
-		super(registry, jobId, jobName, scopeFormat.formatScope(parent, jobId, jobName));
-
-		this.parent = checkNotNull(parent);
+		super(registry, checkNotNull(parent), jobId, jobName, scopeFormat.formatScope(parent, jobId, jobName));
```
shouldn't the `checkNotNull` be performed by the super constructor?

zentol replied:
`parent` is also used in `scopeFormat.formatScope`, as such i wanted the check before that call is made.

tillrohrmann replied:
Ah ok, I see. Looking at the `formatScope` method, shouldn't rather the method enforce the preconditions than the caller?

zentol replied:
Generally i would agree, but this way save a checkNotNull call...

tillrohrmann replied:
Why do you do them at all if you want to save them in the first place?

zentol replied:
I'm only avoiding redundant calls, that's all.

tillrohrmann replied:
But then don't do it here but in the `formatScope` method. Usually the method should enforce that it's preconditions are kept.

zentol replied:
fair enough.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java:
```diff
@@ -0,0 +1,172 @@
+	 * @param name  the metric identifier
+	 * @param value the metric value
+	 */
+	public void add(String name, Object value) {
```
To be honest, I'm not a very big fan of encoding type and hierarchical information in a string which has to be reparsed in order to reconstruct the afore-mentioned information. The problem with this approach is that everything is very implicit and you don't have a tight coupling (in terms of format) between the sender and receiver. If something changes at the sender side, you won't notice at all that you have to change here something as well. Even at runtime the only thing you see is that you don't see the metrics. This makes it very hard to debug. I would be in favour of creating for the different types different messages, e.g. `JobManagerMetric`, `TaskManagerMetric`, `TaskMetric`, etc. These messages contain the respective information. Furthermore, for the histogram type case, you will have to do the whole parsing over and over again instead of doing it once.

zentol replied:
I think we are mixing 2 separate issues here:
A) the format of the transfer
B) how the retrieved metrics are added to the store

Your comment regarding reparsing the string for Histograms targets B. This can be fixed easily by providing specific `addHistogram/...` methods to the store.

Now, let's talk about A. Essentially, the goal was to reduce the overhead for the sender. If you look at the dumping procedure in the `MetricQueryService` you will find that for the sender the current requires the least amount of work, as far as i can tell. We don't create more elaborate containers (which btw. i would personally prefer too!) as this is another short-lived object created for every metric.

Also, in regards to sender/receiver going out of sync: we can add a test to guarantee that sender&receiver do not go out of sync.

tillrohrmann replied:
You're right, I've combined two things here. But B) also depends a little bit on A) (at least implementation wise)

Concerning A) I'm not quite sure whether this is not a classical case of premature optimization. Have we actually measured the impact of the other approach? It would be good to know what it costs us. I would assume that the overhead is negligible.

Furthermore, I'm think you can also achieve an efficient serialization without too many short-living objects if you don't send strings around. First of all, we could serialize all metrics into a continuous byte buffer. Then we wouldn't have the object array overhead. Next we could use our own serializers to store the different object types efficiently (e.g. byte encoding for different metric types). On the server side you would only create the metric name objects once and that's all. Only upon deserialization you would have to create a new set of these objects. But that is ok since it happens on the web server side. The important thing is, though, to have a serialization component which is responsible for serialization and deserialization. That will avoid spreading the encoding over multiple places.

zentol replied:
Wouldn't this just replace the object array overhead with the byte buffer allocation overhead?

tillrohrmann replied:
I think the object array representation (with the contained objects) is not as concise as the byte buffer representation (using your own serialization) because the standard Java serialization will write for every object the class name to identify it.

But again this falls imo into the category of premature optimization.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/messages/Messages.scala:
```diff
@@ -47,4 +47,21 @@ object Messages {
+    *
+    * @return The MetricRequest case object instance.
+    */
+  def getRequestMetrics(): AnyRef = {
```
Can we name this method like the message?

zentol replied:
sure

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1024,6 +1024,17 @@ class JobManager(
+              queryService ! sender()
+            case None =>
+          }
+        case None =>
```
In both `None` cases you should return a failure message to the sender so that it does not have to wait until the future times out.

zentol replied:
If we directly ask the query service we don't need this code anyway.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricQueryService.java:
```diff
@@ -0,0 +1,180 @@
+	 * @return actor reference to the MetricQueryService
+	 */
+	public static ActorRef startMetricQueryService(ActorSystem actorSystem) {
+		return actorSystem.actorOf(Props.create(MetricQueryService.class));
```
Maybe we could give the metric query service a concrete name. Then we would not have to ask the `JobManager` about the metrics. Based on the `JobManager` and `TaskManager` addresses, we could deduce the query metric service's address and ask it directly.

zentol replied:
i see, and how would i do that?

tillrohrmann replied:
`actorSystem.actorOf(Props.create(MetricQueryService.class), "foobar")` creates an actor with the name "foobar". If your actor system is the Flink actor system, then it should be reachable via `akka.tcp://flink@xxx.yyy.zzz:port/foobar` if I'm not mistaken. But you would need an actorsystem reference in the web server to do the actor resolution.

zentol replied:
nvm, may have figured it out...

tillrohrmann commented:
Thanks for your contribution @zentol. I've gone over the code and made some inline comments. My main concern/question is actually the representation of metric's type and hierarchy information. I think that encoding it in a string and then re-parsing it on the receiver side to reconstruct the information is rather fragile and error-prone especially wrt maintainability. Maybe you can give me some background why you decided to do it so.

Apart from that, I think the code contains many tests, which I really like :-)

zentol commented:
@tillrohrmann I've addressed most of your comments. Excluded are calling `checkNotNull` inside `formatScope` and the serialization format.

tillrohrmann commented:
But we still send metric data as strings encoded over the wire and have no checks that the histogram field order is actually correct, right?

zentol commented:
only Gauge values are sent as strings.

tillrohrmann commented:
Sorry, I meant that the hierarchy information is still encoded in a string and then re-parsed. Furthermore, the histogram data is sent as an object array without any information about the field orderings.

zentol commented:
well...currently that is _still_ done. Whether it _will_ be done once this is merged is up in the air.

tillrohrmann commented:
I think this should be addressed (either way) before merging this PR.

zentol commented:
Regarding hierarchy: I'm close to being done with a container for the scope information.

tillrohrmann commented:
Great to hear @zentol 👍

zentol commented:
I've updated and rebased the PR.

The scope information is now stored in a `QueryScopeInfo` inside the `MetricGroups`, for which sub-classes like `TaskQueryScopeInfo` exist. They contain fields for specific values, like the job ID, and the remaining scope not covered by these fields.

Metrics, or rather their scope, name and value(s), are now serialized with a new `MetricDumpSerializer` to a byte[] using a `Data-/ByteArrayOutputStream`. 

On the other end we have the `MetricDumpDeserializer` which deserializes the metrics to a `List<MetricDump>`. A `MetricDump`is a container for the metric value, it's name the `QueryScopeInfo`. There are sub-classes for each metric type, like `CounterDump`.

Neither the `MetricQueryService` nor `MetricFetcher` know anything about the serialized format, just that it's a byte array.

There is no encoding for field orderings but tests that verify that the fields are assigned correctly. If a developer were to change the order of fields a test would fail, and the only way for this to make it into master would be if a) the test is simply changed to give a green light and b) it isn't noticed in the review, at which point all bets are of anyway. So i decided to keep it a bit simpler.

The `MetricStore#addMetric()` method has now become a bit smarter in regards to handling ´Histograms`. With all values being contained in a `HistogramDump` we now only have to analyze the scope once.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java:
```diff
@@ -35,109 +46,111 @@
+			}
+
+			switch (info.getCategory()) {
+				case INFO_CATEGORY_JM:
```
What's the benefit of having an explicit type field over using `instanceof`? I think encoding the type via the actual type has the advantage that you don't mix up classes with wrong category types.

zentol replied:
eh, seemed like the proper way of handling it. Also, (up to) 4 comparisons vs a jump.

tillrohrmann replied:
That is true. Performance-wise it is the more efficient way to execute it, no doubt. I was just wondering whether this is not a case of premature optimization with the price of harder maintainability.

tillrohrmann replied:
On the other hand, it does not seem too overly complicated to be not maintainable. With that in mind, my other comments are mainly obsolete.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java:
```diff
@@ -35,109 +46,111 @@
+
+	private void addMetric(Map<String, Object> target, String name, MetricDump metric) {
+		switch (metric.getCategory()) {
+			case METRIC_CATEGORY_COUNTER:
```
The same question applies here as well. Why don't we simply check what type the metric instance is?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricDump.java:
```diff
@@ -0,0 +1,119 @@
+	 *
+	 * @return category
+	 */
+	public abstract byte getCategory();
```
I think we don't need the explicit category information, because it is already encoded in the sub-types of `MetricDump`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricDump.java:
```diff
@@ -0,0 +1,119 @@
+
+	private MetricDump(QueryScopeInfo scopeInfo, String name) {
+		this.scopeInfo = scopeInfo;
+		this.name = name;
```
`checkNotNull` missing

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricDumpSerialization.java:
```diff
@@ -0,0 +1,269 @@
+		dos.writeDouble(stat.getQuantile(0.999));
+	}
+
+	// ===== Deserialization ===========================================================================================
```
We usually use the following format for structural comments:

```
//-------------------------------------
// Foobar
//-------------------------------------
```

tillrohrmann replied:
Or the dashes replaced with `=`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricDump.java:
```diff
@@ -25,6 +25,7 @@
 	public static final byte METRIC_CATEGORY_COUNTER = 0;
 	public static final byte METRIC_CATEGORY_GAUGE = 1;
 	public static final byte METRIC_CATEGORY_HISTOGRAM = 2;
+	public static final byte METRIC_CATEGORY_METER = 3;
```
No test added for meter metric type.

tillrohrmann replied:
Sorry just saw your latest commit.

tillrohrmann commented:
I think the changes look good. Thanks for your work @zentol :-) I only had a minor question whether we can substitute the explicit category information by the type information of the metric dumps and the `QueryScopeInfo` instances (not for serialization but in the `MetricStore`).

zentol commented:
I'll address the checkNotNull/comment formatting while merging, which I'm doing now. Thank you for looking over it again @tillrohrmann .
